### DISCUSSION TITLE
Is it possible to trigger repository_dispatch by authenticating with GitHub App instead of PAT?

### DISCUSSION THREAD

**Discussion by yukccy** [community/community#101887] [Actions] [answered]
### Select Topic Area

Question

### Body

I would like to add a Jira automation rule that trigger a repository_dispatch job when a Jira issue changed its state. I can successfully trigger the repository_dispatch while using my PAT, but as it is my organization level workflow, it is better not to use my PAT.

The Jira automation rule is actually send a POST request to `https://api.github.com/repos/<owner>/<repo name>/dispatches` with custom data defined `event_type` and `client_payload` and added an `Authorization` header with `Bearer <PAT>` as value.

Do anyone know about authentication by GitHub App or any other ways to achieve same behaviour?

**Comment by ghost** [comment, marked as answer]
Yes, it is possible to authenticate to trigger a repository_dispatch event using a GitHub App instead of a Personal Access Token (PAT). Here are the steps you can follow to set this up:

1. Create a GitHub App:
- Go to your organization's settings, then to Developer settings → GitHub Apps, and create a new GitHub App.
- Set the necessary permissions for the app. For repository_dispatch, you need to have contents: write permission.
- Generate a private key for the GitHub App.

2. Install the GitHub App:
- Install the GitHub App on the repositories where you want to trigger the repository_dispatch event.

3. Obtain a JWT (JSON Web Token):
- Use the private key of the GitHub App to sign a JWT. This JWT is used to authenticate as the GitHub App.

4. Exchange the JWT for an Installation Access Token:
- Send a POST request to the GitHub API to the endpoint /app/installations/:installation_id/access_tokens to get an installation access token. Replace :installation_id with the installation ID of your GitHub App on the repository.

5. Trigger the repository_dispatch Event:
Use the installation access token to authenticate the request to the repository_dispatch endpoint instead of the PAT.

Here's an example of how you might obtain an installation access token using curl and then trigger the repository_dispatch:
```
# Step 3 and 4: Obtain an installation access token
INSTALLATION_ID=<your-installation-id>
JWT=<your-jwt>

ACCESS_TOKEN=$(curl -X POST \
-H "Authorization: Bearer $JWT" \
-H "Accept: application/vnd.github.v3+json" \
https://api.github.com/app/installations/$INSTALLATION_ID/access_tokens | jq -r .token)

# Step 5: Trigger the repository_dispatch
curl -X POST \
-H "Authorization: Bearer $ACCESS_TOKEN" \
-H "Accept: application/vnd.github.v3+json" \
https://api.github.com/repos/<owner>/<repo>/dispatches \
-d '{"event_type": "<event_type>", "client_payload": {"key": "value"}}'
```
Remember to replace <your-installation-id>, <your-jwt>, <owner>, <repo>, and <event_type> with your specific details.

Using a GitHub App for this purpose has several advantages over using a PAT, including finer-grained permissions and not being tied to an individual's GitHub account.

**Reply by yukccy** [reply]
It works on my local machine!

But I am not sure if Jira Automation cannot do all the steps above. It firstly needs to get a JWT from running a Python script, then get the access token by one curl, finally trigger the repository_dispatch event on GitHub by another curl.

Thanks anyway, I will mark it as answer.

**Reply by yukccy** [reply]
Is it possible to get a JWT by POST request?

**Reply by ghost** [reply]
Probably no. In the GitHub App flow, the JWT must be generated on the client side, not through a POST request to GitHub. This is a key part of the authentication process for GitHub Apps. You use the private key of your GitHub App to generate a JWT on your own server or development environment. This JWT includes the app's identifier and a timestamp. With the JWT, you make a POST request to GitHub's API (`/app/installations/:installation_id/access_tokens`) to obtain an installation access token. This POST request uses the JWT for authentication. Finally, you use the obtained installation access token to authenticate a request to the `repository_dispatch` endpoint of the GitHub API. In this process, the JWT is not obtained via a POST request but is created using your own tooling (like a script or software library) that can generate JWTs. This is a fundamental part of the GitHub App authentication mechanism, which differs from typical OAuth flows where tokens are directly obtained from the server via POST requests.

**Reply by yukccy** [reply]
Make sense. I will try to find way to get JWT token for Jira. Thanks!
